### PR TITLE
Pv2 period end pv2 1194 pv2 1196 pv2 1197 pv2 1198 pv2 1207

### PR DESCRIPTION
**Pull request by abroodie**

### CODE REVIEW

File: src/SFA.DAS.Payments.Database/zzz.Script.PostDeployment.sql [hunk 1]
```diff
@@ -40,7 +40,9 @@ IF NOT EXISTS (SELECT * FROM [Payments2].[JobType]  WHERE [Id] = 1)
 GO 
 
 IF NOT EXISTS (SELECT * FROM [Payments2].[JobType]  WHERE [Id] = 2)
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.Database/zzz.Script.PostDeployment.sql]: This should be changed to use Merge Statement

File: src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/JobContextMessageHandlerTests.cs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+                    { JobContextMessageConstants.KeyValuePairs.FundingFm36Output, "invalid path" },
+                    { JobContextMessageConstants.KeyValuePairs.FundingFm36OutputPeriodEnd, "valid path" },
+                    { JobContextMessageConstants.KeyValuePairs.Container, "container" },
+                }
+            };
+
+            var handler = mocker.Create<JobContextMessageHandler>();
+            await handler.HandleAsync(jobContextMessage, CancellationToken.None);
+
+            mocker.Mock<IFileService>().Verify(svc => svc.OpenReadStreamAsync(
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/JobContextMessageHandlerTests.cs]: Can we have a TearDown method that ensures that all expected mocks are verified?

File: src/SFA.DAS.Payments.EarningEvents.Application/Handlers/JobContextMessageHandler.cs [hunk 3]
```diff
@@ -119,11 +126,13 @@ private async Task SendReceivedEarningsEvent(long jobId, DateTime ilrSubmissionD
             await endpointInstance.Publish(message).ConfigureAwait(false);
         }
 
-        private async Task<FM36Global> GetFm36Global(JobContextMessage message, int collectionPeriod, CancellationToken cancellationToken )
+        private async Task<FM36Global> GetFm36Global(JobContextMessage message, int collectionPeriod, CancellationToken cancellationToken)
         {
             FM36Global fm36Output;
-            var fileReference = message.KeyValuePairs[JobContextMessageKey.FundingFm36Output].ToString();
-            var container = message.KeyValuePairs[JobContextMessageKey.Container].ToString();
+            var fileReference = message.Topics.Any(topic => topic.Tasks.Any(task => task.Tasks.Any(taskName => taskName.Equals(JobContextMessageConstants.Tasks.ProcessPeriodEndSubmission))))
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.EarningEvents.Application/Handlers/JobContextMessageHandler.cs]: This doesn't seem right to me  as here we are only interested in FundingFm36Output file so we should ignore any file reference aside from "FundingFm36Output"

File: src/SFA.DAS.Payments.EarningEvents.Application/Infrastructure/Ioc/EarningEventsModule.cs [hunk 4]
```diff
@@ -51,6 +54,8 @@ protected override void Load(ContainerBuilder builder)
             //    c.Resolve<IPaymentLogger>(),
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.EarningEvents.Application/Infrastructure/Ioc/EarningEventsModule.cs]: Can you please remove comments please

File: src/SFA.DAS.Payments.EarningEvents.ServiceFabric/ApplicationPackageRoot/ApplicationManifest.xml [hunk 5]
```diff
@@ -44,6 +46,8 @@
             <Parameter Name="ApplicationInsightsInstrumentationKey" Value="[ApplicationInsightsInstrumentationKey]" />
             <Parameter Name="LogLevel" Value="[LogLevel]" />
             <Parameter Name="ProcessMessageSequentially" Value="[ProcessMessageSequentially]" />
+            <Parameter Name="TimeToPauseBetweenChecks" Value="[TimeToPauseBetweenChecks]" />
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.EarningEvents.ServiceFabric/ApplicationPackageRoot/ApplicationManifest.xml]: This config should be added to the cloud.xml file so we can control it from  VSTS

File: src/SFA.DAS.Payments.FundingSource.Application/Services/PeriodEndService.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+        public PeriodEndService(ILevyFundingSourceRepository repository, IPaymentLogger logger)
+        {
+            this.repository = repository ?? throw new ArgumentNullException(nameof(repository));
+            this.logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public async Task<List<ProcessLevyPaymentsOnMonthEndCommand>> GenerateEmployerPeriodEndCommands(PeriodEndRunningEvent message)
+        {
+            logger.LogInfo($"Sending requests to trigger month end for job: {message.JobId}");
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.FundingSource.Application/Services/PeriodEndService.cs]: Can you please reference  ESFA.DC.Logging  Nuget Package I am getting a warning message in code.

File: src/SFA.DAS.Payments.FundingSource.Application/Repositories/LevyFundingSourceRepository.cs [hunk 7]
```diff
@@ -50,5 +50,22 @@ await dataContext.EmployerProviderPriority
...
         }
 
+        public async Task<List<long>> GetEmployerAccounts(CancellationToken cancellationToken)
+        {
+            var transferSenders = (await dataContext.Apprenticeship
+                .Where(apprenticeship => apprenticeship.TransferSendingEmployerAccountId != null && apprenticeship.TransferSendingEmployerAccountId != 0)
+                .Select(apprenticeship => apprenticeship.TransferSendingEmployerAccountId)
+                .Distinct()
+                .ToListAsync(cancellationToken))
+                .Select(accountId => accountId.Value)
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.FundingSource.Application/Repositories/LevyFundingSourceRepository.cs]: No need for this code and another ToList, just select in Line 56 just select the Value.

File: src/SFA.DAS.Payments.FundingSource.LevyFundedProxyService/Handlers/PeriodEndHandler.cs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+            this.logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            this.periodEndService = periodEndService ?? throw new ArgumentNullException(nameof(periodEndService));
+        }
+
+        public async Task Handle(PeriodEndRunningEvent message, IMessageHandlerContext context)
+        {
+            try
+            {
+                logger.LogInfo($"Received Period End event. Now getting employer period end commands for collection: {message.CollectionPeriod.Period:00}-{message.CollectionPeriod.AcademicYear:0000}.");
+                var employerPeriodEndCommands = await periodEndService.GenerateEmployerPeriodEndCommands(message);
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.FundingSource.LevyFundedProxyService/Handlers/PeriodEndHandler.cs]: Please add ConfigureAwait

File: src/SFA.DAS.Payments.JobContextMessageHandling/JobContextManagerService.cs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+            this.logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            this.jobContextManager = jobContextManager ?? throw new ArgumentNullException(nameof(jobContextManager));
+        }
+
+        public async Task RunAsync(CancellationToken cancellationToken)
+        {
+            var initialised = false;
+            try
+            {
+                logger.LogDebug("Starting the Earning Events service.");
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.JobContextMessageHandling/JobContextManagerService.cs]: This Logging comment doesn't seem right can you change this to -  "Starting DC job context manager"

File: src/SFA.DAS.Payments.JobContextMessageHandling/JobContextManagerService.cs [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+                jobContextManager.OpenAsync(cancellationToken);
+                initialised = true;
+                logger.LogInfo("Started the Period End service.");
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.JobContextMessageHandling/JobContextManagerService.cs]: This Logging comment doesn't seem right can you change this to -  "Started DC job context manager"

File: src/SFA.DAS.Payments.PeriodEnd.Messages.Internal/Class1.cs [hunk 11]
```diff
@@ -0,0 +1,8 @@
+﻿using System;
+
+namespace SFA.DAS.Payments.PeriodEnd.Messages.Internal
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.PeriodEnd.Messages.Internal/Class1.cs]: Please remove this file.

File: src/SFA.DAS.Payments.PeriodEnd.PeriodEndService/PeriodEndService.cs [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+using Autofac;
+using Microsoft.ServiceFabric.Services.Communication.Runtime;
+using Microsoft.ServiceFabric.Services.Runtime;
+using SFA.DAS.Payments.Application.Infrastructure.Logging;
+using SFA.DAS.Payments.JobContextMessageHandling;
+using SFA.DAS.Payments.ServiceFabric.Core;
+
+namespace SFA.DAS.Payments.PeriodEnd.PeriodEndService
+{
+    /// <summary>
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.PeriodEnd.PeriodEndService/PeriodEndService.cs]: remove comments please

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/PeriodEndService.cs [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+        public PeriodEndService(IPaymentLogger  logger, IProviderPaymentsRepository repository)
+        {
+            this.logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            this.repository = repository ?? throw new ArgumentNullException(nameof(repository));
+        }
+
+        public async Task<List<ProcessProviderMonthEndCommand>> GenerateProviderMonthEndCommands(PeriodEndStoppedEvent message)
+        {
+            logger.LogInfo($"now building Provider Period End Commands for collection period: {message.CollectionPeriod.Period:00}-{message.CollectionPeriod.AcademicYear:0000}, job: {message.JobId}");
+            var providers = await repository.GetMonthEndProviders(message.CollectionPeriod, CancellationToken.None)
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.ProviderPayments.Application/Services/PeriodEndService.cs]: I noticed the repository is not selecting distinct UKPRN, so we can be sending - ProcessProviderMonthEndCommand multiple times for the same provider. Please fix the repository query to distinct on the UKPRN.

File: src/SFA.DAS.Payments.Database/zzz.Script.PostDeployment.sql (hunk 1 above)
Comment (abroodie) [reply src/SFA.DAS.Payments.Database/zzz.Script.PostDeployment.sql]: No, this code works fine.

File: src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/JobContextMessageHandlerTests.cs (hunk 2 above)
Comment (abroodie) [reply src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/JobContextMessageHandlerTests.cs]: no, this tests the things that need to be tested and makes it obvious what is being tested.

File: src/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService/RequiredPaymentsService.cs [hunk 14]
```diff
@@ -147,12 +160,44 @@ protected override async Task OnActivateAsync()
...
+        }
+
+        private async Task StoreCollectionPeriod(CollectionPeriod collectionPeriod)
+        {
+            await collectionPeriodCache.AddOrReplace("CollectionPeriod", collectionPeriod).ConfigureAwait(false);
+        }
+
         public async Task Initialise()
         {
             if (await StateManager.ContainsStateAsync(CacheKeys.InitialisedKey).ConfigureAwait(false)) return;
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService/RequiredPaymentsService.cs]: You should use ActorDataCache for this

File: src/SFA.DAS.Payments.EarningEvents.Application/Handlers/JobContextMessageHandler.cs (hunk 3 above)
Comment (abroodie) [reply src/SFA.DAS.Payments.EarningEvents.Application/Handlers/JobContextMessageHandler.cs]: The file reference is different if the submission is a period end submisson.

File: src/SFA.DAS.Payments.EarningEvents.Application/Infrastructure/Ioc/EarningEventsModule.cs (hunk 4 above)
Comment (abroodie) [reply src/SFA.DAS.Payments.EarningEvents.Application/Infrastructure/Ioc/EarningEventsModule.cs]: Done

File: src/SFA.DAS.Payments.EarningEvents.ServiceFabric/ApplicationPackageRoot/ApplicationManifest.xml (hunk 5 above)
Comment (abroodie) [reply src/SFA.DAS.Payments.EarningEvents.ServiceFabric/ApplicationPackageRoot/ApplicationManifest.xml]: done

File: src/SFA.DAS.Payments.FundingSource.Application/Services/PeriodEndService.cs (hunk 6 above)
Comment (abroodie) [reply src/SFA.DAS.Payments.FundingSource.Application/Services/PeriodEndService.cs]: done

File: src/SFA.DAS.Payments.FundingSource.Application/Repositories/LevyFundingSourceRepository.cs (hunk 7 above)
Comment (abroodie) [reply src/SFA.DAS.Payments.FundingSource.Application/Repositories/LevyFundingSourceRepository.cs]: Feels like you are making sql  do more work just reduce the number of lines of code.

File: src/SFA.DAS.Payments.FundingSource.LevyFundedProxyService/Handlers/PeriodEndHandler.cs (hunk 8 above)
Comment (abroodie) [reply src/SFA.DAS.Payments.FundingSource.LevyFundedProxyService/Handlers/PeriodEndHandler.cs]: done

File: src/SFA.DAS.Payments.JobContextMessageHandling/JobContextManagerService.cs (hunk 9 above)
Comment (abroodie) [reply src/SFA.DAS.Payments.JobContextMessageHandling/JobContextManagerService.cs]: done

File: src/SFA.DAS.Payments.JobContextMessageHandling/JobContextManagerService.cs (hunk 10 above)
Comment (abroodie) [reply src/SFA.DAS.Payments.JobContextMessageHandling/JobContextManagerService.cs]: done

File: src/SFA.DAS.Payments.PeriodEnd.PeriodEndService/PeriodEndService.cs (hunk 12 above)
Comment (abroodie) [reply src/SFA.DAS.Payments.PeriodEnd.PeriodEndService/PeriodEndService.cs]: done

File: src/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService/RequiredPaymentsService.cs (hunk 14 above)
Comment (abroodie) [reply src/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService/RequiredPaymentsService.cs]: create a separate ticket for this

File: src/SFA.DAS.Payments.PeriodEnd.Messages.Internal/Class1.cs (hunk 11 above)
Comment (abroodie) [reply src/SFA.DAS.Payments.PeriodEnd.Messages.Internal/Class1.cs]: done

File: src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/JobContextMessageHandlerTests.cs (hunk 2 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/JobContextMessageHandlerTests.cs]: Is not the main concern so leaving it

File: src/SFA.DAS.Payments.FundingSource.Application/Repositories/LevyFundingSourceRepository.cs (hunk 7 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.FundingSource.Application/Repositories/LevyFundingSourceRepository.cs]: Sql will not do more work as we have already filtered the result to only bring TransferSendingEmployerAccountId that has value

File: src/SFA.DAS.Payments.FundingSource.LevyFundedProxyService/Handlers/PeriodEndHandler.cs (hunk 8 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.FundingSource.LevyFundedProxyService/Handlers/PeriodEndHandler.cs]: still, no configure await

File: src/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService/RequiredPaymentsService.cs (hunk 14 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService/RequiredPaymentsService.cs]: Can change it later

**Pull request merged**